### PR TITLE
Add Squeezeformer to ASR

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Adds SqueezeFormer model to ASR collections

**Collection**: [ASR]

# Changelog 
- Adds prototype configs, code and model level support for Squeezeformer models
- Checkpoints may be provided at a later date after debugging smaller sized models.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 1]
```diff
@@ -137,41 +138,60 @@ class ConformerConvolution(nn.Module):
     Args:
         d_model (int): hidden dimension
         kernel_size (int): kernel size for depthwise convolution
+        pointwise_activation (str): name of the activation function to be used for the pointwise conv.
+            Note that Conformer uses a special key `glu_` which is treated as the original default from
+            the paper.
     """
 
-    def __init__(self, d_model, kernel_size, norm_type='batch_norm'):
+    def __init__(self, d_model, kernel_size, norm_type='batch_norm', pointwise_activation='glu_'):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:146]: Why it is not just glu?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 2]
```diff
@@ -190,6 +210,19 @@ def forward(self, x, pad_mask=None):
         x = x.transpose(1, 2)
         return x
 
+    def reset_parameters_conv(self):
+        pw1_max = self.d_model ** -0.5
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:214]: What are there two pw1_max and pw2_max?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 3]
```diff
@@ -190,6 +210,19 @@ def forward(self, x, pad_mask=None):
         x = x.transpose(1, 2)
         return x
 
+    def reset_parameters_conv(self):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:213]: Isn't it better to call the method just reset_parameters to be consistent?
The same for the FF module.

File: nemo/collections/asr/parts/submodules/squeezeformer_modules.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+from nemo.collections.asr.parts.submodules.conformer_modules import ConformerConvolution, ConformerFeedForward
+from nemo.collections.asr.parts.submodules.multi_head_attention import (
+    MultiHeadAttention,
+    RelPositionMultiHeadAttention,
+)
+from nemo.collections.asr.parts.utils.activations import Swish
+from nemo.core.classes.mixins import AccessMixin
+from nemo.core.classes.mixins.adapter_mixins import AdapterModuleMixin
+from nemo.utils import logging
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/squeezeformer_modules.py:27]: Logging and Swish are not used in this file.

File: nemo/core/optim/lr_scheduler.py [hunk 5]
```diff
@@ -493,6 +502,52 @@ def _noam_annealing(self, initial_lr, step):
         return out_lr
 
 
+class NoamHoldAnnealing(WarmupHoldPolicy):
+    def __init__(self, optimizer, *, max_steps, decay_rate=0.5, min_lr=0.0, last_epoch=-1, **kwargs):
+        """
+        Implementation of the Noam Hold Annealing policy from the SqueezeFormer paper.
```
Comment (VahidooX) [inline nemo/core/optim/lr_scheduler.py:508]: Should we have more detail on how it works?

File: nemo/collections/asr/modules/squeezeformer_encoder.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+from nemo.collections.asr.parts.submodules.multi_head_attention import PositionalEncoding, RelPositionalEncoding
+from nemo.collections.asr.parts.submodules.squeezeformer_modules import SqueezeformerLayer
+from nemo.collections.asr.parts.submodules.subsampling import ConvSubsampling, StackingSubsampling, TimeReductionModule
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.exportable import Exportable
+from nemo.core.classes.mixins import adapter_mixins
+from nemo.core.classes.module import NeuralModule
+from nemo.core.neural_types import AcousticEncodedRepresentation, LengthsType, NeuralType, SpectrogramType
+from nemo.utils import logging
```
Comment (VahidooX) [inline nemo/collections/asr/modules/squeezeformer_encoder.py:31]: logging not used here.

File: nemo/collections/asr/modules/squeezeformer_encoder.py [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+from nemo.utils.decorators import experimental
+
+__all__ = ['SqueezeformerEncoder']
+
+
+@experimental
+class SqueezeformerEncoder(NeuralModule, Exportable):
```
Comment (VahidooX) [inline nemo/collections/asr/modules/squeezeformer_encoder.py:38]: Let's add the Squeezformer to the readme page of repo.

File: examples/asr/conf/squeezeformer/squeezeformer_ctc_bpe.yaml [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    n_layers: 18
+    d_model: 512
+
+    # Squeezeformer params
+    adaptive_scale: true
+    time_reduce_idx: 8
+    time_recovery_idx: null
+
+    # Sub-sampling params
+    subsampling: dw_striding # vggnet, striding or dw_striding, vggnet may give better results but needs more memory
```
Comment (VahidooX) [inline examples/asr/conf/squeezeformer/squeezeformer_ctc_bpe.yaml:108]: stacking is also an option.

**Review by VahidooX — CHANGES_REQUESTED** [review]
LGTM! Just minor comments.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:146]: So the string "glu" can be activation without the doubling of filters in the dw_conv_input_dim, but for backward compact we want to support "glu" but without the double channel (original conformed) - so _ is to mark it as the default used by conformed (originally it cant be modified by config so it should be safe).

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 9]
```diff
@@ -137,41 +138,60 @@ class ConformerConvolution(nn.Module):
...
+        self.kernel_size = kernel_size
+
+        if pointwise_activation in activation_registry:
+            self.pointwise_activation = activation_registry[pointwise_activation]()
+            dw_conv_input_dim = d_model * 2
+
+            if hasattr(self.pointwise_activation, 'inplace'):
+                self.pointwise_activation.inplace = True
+        else:
+            self.pointwise_activation = pointwise_activation
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:159]: Its not always necessary to be GLU though, so best keep it as is.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/conformer_modules.py]: Just named for the two pointwise ops, can be merged into one line. Done.

File: nemo/core/optim/lr_scheduler.py (hunk 5 above)
Comment (titu1994) [reply nemo/core/optim/lr_scheduler.py:508]: Yes good point.

**Review by VahidooX — APPROVED** [review]
LGTM!

**Review by VahidooX — APPROVED** [review]

**Pull request merged**